### PR TITLE
make metrics backwards compatible

### PR DESCRIPTION
**Pull request by FZambia**

### CODE REVIEW

File: libcentrifugo/metrics.go [hunk 1]
```diff
@@ -120,7 +101,7 @@ func (c *metricCounter) LoadRaw() int64 {
 }
 
 func (c *metricCounter) LastIn() int64 {
-	return atomic.LoadInt64(&c.value)
+	return atomic.LoadInt64(&c.lastIntervalDelta)
```
Comment (banks) [inline libcentrifugo/metrics.go]: Why atomic load? Before all reads and writes to delta were done only under lock. But I may be missing something on my phone!

File: libcentrifugo/metrics.go [hunk 2]
```diff
@@ -19,48 +18,66 @@ import (
...
+	CPU int64 `json:"cpu_usage"`
+}
 
+// metricsRegistry contains various Centrifugo statistic and metric information aggregated
+// once in a configurable interval.
+type metricsRegistry struct {
 	// mu protects from multiple processes updating snapshot values at once
 	// but raw counters may still increment atomically while held so it's not a strict
 	// point-in-time snapshot of all values.
 	mu sync.Mutex
```
Comment (banks) [inline libcentrifugo/metrics.go]: This unaligns all the counters below. You said "no need to align" but I don't see why not - as long as you are still dong atomic ops on the ints in the struct a below they will still need to be aligned. 

Perhaps a murex happens to be 8 bytes? Even if so (and my gut feeling is it isn't) that is a platform/implementation detail I don't think is guaranteed so we shouldn't rely on it. 

I think since the bool is gone we can just move this to the end. 

We probably need to include in comments on both these structs reminders not to unaligns them again.

**Comment by banks** [comment]
In general this is a lot better and what I should have done. I was really trying to limit how many places copy-pasted the big list of counter fields but that's a dumb think to complicate code for and still didn't hug my goal of just one definition. 

See online though. Unless I'm missing something (which is quite likely - it's nearly 3am and I'm on my phone while pacifying a baby) this doesn't fix the issue of unaligned atomic 64 ops on the counters due to mutex. 

Incidentally did you try to run 1.4.0 i386 to see if you could duplicate? Not sure where to get hold of a real 32 bit machine to try on but it might be enough to use 32-bit OS in a vm on your Mac or on Digital Ocean etc.

**Comment by banks** [comment]
From sync docs: 

> ☞ On x86-32, the 64-bit functions use instructions unavailable before the Pentium MMX. On non-Linux ARM, the 64-bit functions use instructions unavailable before the ARMv6k core. On both ARM and x86-32, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a global variable or in an allocated struct or slice can be relied upon to be 64-bit aligned.

So same bug would happen on any ARM processor too it seems. You could probably reproduce on http://www.scaleway.com server for a few cents if no luck getting an i386 vm to fail.

**Comment by FZambia** [comment]
That struct does not contain int64 values we access in atomic way, should it still be aligned? Will try to reproduce panic on 32-bit machine now...

**Comment by FZambia** [comment]
@banks just tried on Centos 6 32-bit. Reproduced panic - happens after every request to API. No problems with version built from this branch (btw still panics in current master branch - so the problem is really in aligning).

I understand that your code was a bit more generic in the end. But with this approach we preserve the same exported structs as before, so `gocent` continues to work. And this approach is a bit more simple to understand in my opinion because we don't need to have custom marshalers and clone counters.

**Comment by FZambia** [comment]
Loading deltas atomically seems really non needed - will fix right now

**Comment by banks** [comment]
> That struct does not contain int64 values we access in atomic way

Yes it does since the metricCounter fields are not pointers. That means that each field of the metricCounter is laid out continuously in memory inside the MetricsRegisty and so will be offset by the mutex. 

It's the same as before - the bool was at the end so it wasn't causing align issues with the counters in the _same_ metricCounter, but rather causing all the subsequent ones in the Metrics struct to be offset by one byte each time. 

If this works now in 32 bit then it means that sync.Mutex just happens to be 64 bits on your architecture. But that is not a good thing to rely on - on a genuinene 32 bit processor it would almost certainly be 32 bits (guess though).

At any rate the documentation from sync/atomic is very clear: the _only_ layout guarantees you can rely on (without using unsafe.SizeOf or similar) are that structs are allocated 8-byte aligned. 

There are two options I see:
 - move the mutex to the end
 - change all the fields to *metricCounter/Gauge so they are all allocated separately and thus all guaranteed to individually align.

And we should comment as I mentioned so we don't accidentally reintroduce align errors if we change this in future.

**Comment by FZambia** [comment]
hm, I have not thought about this. ok, let's move it in the end and comment then.

**Comment by banks** [comment]
Turns out I'm not 100% accurate with my assumptions

`sync.Mutex` is [defined in current implementation](https://golang.org/src/sync/mutex.go?s=729:777#L12) like this:

```go
type Mutex struct {
    state int32
    sema  uint32
}
```

Which means it's size will not be architecture dependent and we can rely on this PR working _for the current sync.Mutex implementation_.

So it's not a bug right now, but is a potential future bug since we are implicitly relying on the `sync.Mutex` implementation to never changed in a future go version.

**tl;dr** this is fine for now and fixes issues. We _should_ change it though so that we are not relying on undocumented implementation details of mutex to be correct. I'll make a separate PR.
